building/glass/inner.py

```diff
@@ -0,0 +1,612 @@
+            # filesystem, probably because the build process needs too much space
+            tempdir = ALTERNATE_TEMPDIR_LOCATION
+            if not os.path.isdir(tempdir):
+                os.makedirs(tempdir)
+            log("glass", "using alternate tempdir location", tempdir)
+
+        with tempfile.TemporaryDirectory(prefix="glass-", dir=tempdir) as d:
+            exc = Execution(self, os.path.join(d, "stage"), os.path.join(d, "tree"), branch, tempdir=tempdir)
+            for step in self.glassfile["build"].array():
+                exc.add_parse(step)
+            if self.build_type == "deb":
+                depends = [dep.string() for dep in self.glassfile["control"].get("depends").array(True)]
+                install_scripts = self.glassfile["control"].get("install-scripts", {})
+                debuild.validate_install_script_names(install_scripts.keys())
+                install_scripts = {key: os.path.join(self.path, value.string()) for key, value in install_scripts.items()}
+                exc.add_step(lambda: exc.perform_final_debuild(self.change_date, depends, install_scripts))
+            elif self.build_type == "aci":
+                set_exec = self.glassfile["control"].get("set-exec").string(True)
+                ports = self.glassfile["control"].get("ports").array(True)
+                real_ports = []
+                for port in ports:
+                    pname, protocol, pnum = check_parameters(port, ("name", "protocol", "port"))
+                    real_ports.append((pname.string(), protocol.string(), pnum.integer()))
+                exc.add_step(lambda: exc.perform_final_acbuild(self.change_date, set_exec, real_ports))
+            elif self.build_type == "tgz":
+                exc.add_step(lambda: exc.perform_final_tarball())
+            else:
+                raise Exception("only implemented for debian packages currently")
+            exc.run(debug=debug)
+
+        if self.is_built(branch):
+            log("glass", "package", self.package_name, "built for", branch)
+            log("glass", "*********************")
+        else:
+            raise Exception("internal error: package failed to build: %s" % self.package_name)
+
+
+def log(operation, *message):
+    print(("[%s]" % operation).rjust(13), " ".join(map(str, message)))
+
+
+class Execution:
+    def __init__(self, project: Project, stagedir: str, outputdir: str, branch: str, tempdir: str):
+        self.project = project
+        self.stagedir = stagedir
+        self.outputdir = outputdir
+        self.tempdir = tempdir
+        self.steps = []
+        self.branch = branch
+
+    def add_step(self, step_func):
+        self.steps.append(step_func)
+
+    # returns a function that can be called with the build tree
+    def add_parse(self, step_yaml: yamlobj.YamlObject):
+        step_type = step_yaml["type"].string()
+        if step_type == "upstream-extract":
```

Comment:
abstract the branches of this giant if/else

Reply:
Good point. Might not get to that right now.